external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala

```diff
@@ -0,0 +1,143 @@
+
+  import testImplicits._
+
+  private var testUtils: KafkaTestUtils = _
+  private var kafkaParams: Map[String, String] = _
+
+  override val streamingTimout = 10.seconds
+
+  case class AddKafkaData(
+      kafkaSource: KafkaSource,
+      topic: String, data: Int*)(implicit multiPartitionCheck: Boolean = true) extends AddData {
+
+    override def addData(): Offset = {
+      val sentMetadata = testUtils.sendMessages(topic, data.map{ _.toString}.toArray)
+      val latestOffsetMap = sentMetadata
+        .groupBy { _._2.partition }
+        .map { case (partition, msgAndMetadata) =>
+          val maxOffsetInPartition = msgAndMetadata.map(_._2).maxBy(_.offset).offset()
+          (TopicAndPartition(topic, partition), maxOffsetInPartition + 1)
+        }
+
+      def metadataToStr(m: (String, RecordMetadata)): String = {
+        s"Sent ${m._1} to partition ${m._2.partition()}, offset ${m._2.offset()}"
+      }
+
+      assert(latestOffsetMap.size > 1,
+        s"Added data does not test multiple partitions: " + sentMetadata.map(metadataToStr))
+
+      // Expected offset to ensure this data is read is last offset of this data + 1
+      val offset = KafkaSourceOffset(latestOffsetMap)
+      logInfo(s"Added data, expected offset $offset")
+      offset
+    }
+
+    override def source: Source = kafkaSource
+  }
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    testUtils = new KafkaTestUtils
+    testUtils.setup()
+    kafkaParams = Map("metadata.broker.list" -> testUtils.brokerAddress)
+  }
+
+  override def afterAll(): Unit = {
+    if (testUtils != null) {
+      testUtils.teardown()
+      testUtils = null
+      super.afterAll()
+    }
+  }
+
+  test("basic receiving from latest offset with 1 topic and 2 partitions") {
+    val topic = "topic1"
+    testUtils.createTopic(topic, partitions = 2)
+
+    // Add data in multiple rounds to the same topic and test whether
+    for (i <- 0 until 5) {
+      logInfo(s"Round $i")
+
+      // Create Kafka source that reads from latest offset
+      val kafkaSource = KafkaSource(Set(topic), kafkaParams)
+
+      val mapped =
+        kafkaSource
+          .toDS()
+          .map[Int]((kv: (Array[Byte], Array[Byte])) => new String(kv._2).toInt + 1)
+
+      logInfo(s"Initial offsets: ${kafkaSource.initialOffsets}")
+
+      testStream(mapped)(
+        AddKafkaData(kafkaSource, topic, 1, 2, 3),
+        CheckAnswer(2, 3, 4),
+        StopStream ,
+        DropBatches(1),         // Lose last batch in the sink
+        StartStream,
+        CheckAnswer(2, 3, 4),   // Should get the data back on recovery
+        StopStream,
+        AddKafkaData(kafkaSource, topic, 4, 5, 6),    // Add data when stream is stopped
+        StartStream,
+        CheckAnswer(2, 3, 4, 5, 6, 7),                // Should get the added data
+        AddKafkaData(kafkaSource, topic, 7, 8),
+        CheckAnswer(2, 3, 4, 5, 6, 7, 8, 9)
+      )
+    }
+  }
+
+  compare(
```

Comment:
I would put these in their own suite since they are proper unit tests (while the other seem more like integration tests).